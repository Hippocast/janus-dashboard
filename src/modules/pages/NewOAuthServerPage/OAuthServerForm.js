import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Field, formValueSelector, reduxForm } from 'redux-form';

import PLACEHOLDER from '../../../configurations/placeholders.config';

import transformFormValues from '../../../helpers/transformFormValues';
import block from '../../../helpers/bem-cn';
import checkOnPattern from '../../../helpers/pattern-check';

import Section from '../../Layout/Section/Section';
import Row from '../../Layout/Row/Row';
import Title from '../../Layout/Title/Title';
import Input from '../../inputs/Input';
import Radio from '../../inputs/Radio/Radio';
import Label from '../../labels/Label';
import Hint from '../../labels/Hint/Hint';
import MultiSelect from '../../selects/MultiSelect/MultiSelect';
import TagSelect from '../../selects/TagSelect/TagSelect';

import Button from '../../buttons/Button';
import RenderPlugins from '../../forms/plugins/RenderPlugins';

// import './NewApiForm.css';

const b = block('j-api-form');
const col = block('j-col');

const propTypes = {
    schema: PropTypes.object.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    // initialValues: PropTypes.object,
};

const OAuthServerForm = props => {
    const {
        handleSubmit,
        schema,

        apiSchema,
        includePlugin,
        removePlugin,
        initialValues,
        plugins,
        excludePlugin,
        selectPlugin,
        selectedPlugins,
    } = props;
    const parse = value => value && parseInt(value);
    const optionsTransformer = config => config.map(item => ({
        label: item,
        value: item,
    }));

    return (
        <form className={b} onSubmit={handleSubmit}>
            <Section>
                <Row>
                    <Title>Create New oAuth Server</Title>
                </Row>
            </Section>
            <div className={b('inner')}>
                <div className={b('section')}>
                    <div className={b('section-title')}>1. General</div>
                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>oAuth Server Name</Label>
                            <Field
                                name="name"
                                type="text"
                                component={Input}
                            />
                            <Hint>Must be unique</Hint>
                        </Row>
                    </Row>
                </div>

                <div className={b('section')}>
                    <div className={b('section-title')}>2. Cors_meta</div>

                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Is Enabled?</Label>
                            <Row className={b('radio-wrap')()}>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="cors_meta.active"
                                        component={Radio}
                                        value={'true'}
                                        type="radio"
                                        id="is-active"
                                    />
                                    <Label htmlFor="is-active">Yes</Label>
                                </Row>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="cors_meta.active"
                                        component={Radio}
                                        value={'false'}
                                        type="radio"
                                        id="is-not-active"
                                    />
                                    <Label htmlFor="is-not-active">No</Label>
                                </Row>
                            </Row>
                        </Row>
                        {/*
                        <Row col>
                            <Label>Domains</Label>
                            <Field
                                name="cors_meta.domains"
                                type="text"
                                placeholder="*"
                                component={Input}
                            />
                            <Hint>A list of all domains from which the endpoint will accept requests</Hint>
                        </Row>
                        */}
                    </Row>

                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Domains</Label>
                            <Field
                                name="cors_meta.domains"
                                type="text"
                                placeholder={PLACEHOLDER.DOMAINS}
                                component={Input}
                            />
                            <Hint>A list of all domains from which the endpoint will accept requests</Hint>
                        </Row>
                        <Row col>
                            <Label>Methods</Label>
                            <Field
                                name="cors_meta.methods"
                                type="text"
                                edit={false}
                                value="cors_meta.methods"
                                options={optionsTransformer(schema.cors_meta.methods)}
                                component={MultiSelect}
                            />
                            <Hint>HTTP methods that are supported for the endpoint.</Hint>
                        </Row>
                    </Row>

                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Request Headers</Label>
                            <Field
                                name="cors_meta.request_headers"
                                type="text"
                                edit={false}
                                value="cors_meta.request_headers"
                                placeholder={PLACEHOLDER.REQUEST_HEADERS}
                                options={optionsTransformer(schema.cors_meta.request_headers)}
                                component={TagSelect}
                            />
                            <Hint>Value(s) for the Access-Control-Allow-Headers header.</Hint>
                        </Row>
                        <Row col>
                            <Label>Exposed Headers</Label>
                            <Field
                                name="cors_meta.exposed_headers"
                                type="text"
                                edit={false}
                                value="cors_meta.exposed_headers"
                                placeholder={PLACEHOLDER.EXPOSED_HEADERS}
                                options={optionsTransformer(schema.cors_meta.exposed_headers)}
                                component={TagSelect}
                            />
                            <Hint>Value for the Access-Control-Expose-Headers header.</Hint>
                        </Row>
                    </Row>

                    {/*
                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Methods</Label>
                            <Field
                                name="proxy.methods"
                                type="text"
                                placeholder="Choose one or more methods"
                                value={[]}
                                options={optionsTransformer(initialValues.proxy.methods)}
                                component={MultiSelect}
                            />
                            <Hint>HTTP methods that are supported for the endpoint.</Hint>
                        </Row>
                        <Row col>
                            <Label>Preserve Host?</Label>
                            <Row className={b('radio-wrap')()}>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.preserve_host"
                                        component={Radio}
                                        value={'true'}
                                        type="radio"
                                        id="preserve-host-true"
                                    />
                                    <Label htmlFor="preserve-host-true">Yes</Label>
                                </Row>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.preserve_host"
                                        component={Radio}
                                        value={'false'}
                                        type="radio"
                                        id="preserve-host-false"
                                    />
                                    <Label htmlFor="preserve-host-false">No</Label>
                                </Row>
                            </Row>
                            <Hint>Preserve the host header the client used for the incoming request.</Hint>
                        </Row>
                    </Row>
                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Append Path?</Label>
                            <Row className={b('radio-wrap')()}>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.append_path"
                                        component={Radio}
                                        value={'true'}
                                        type="radio"
                                        id="append-path-true"
                                    />
                                    <Label htmlFor="append-path-true">Yes</Label>
                                </Row>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.append_path"
                                        component={Radio}
                                        value={'false'}
                                        type="radio"
                                        id="append-path-false"
                                    />
                                    <Label htmlFor="append-path-false">No</Label>
                                </Row>
                            </Row>
                            <Hint>Appends the path from the listen_path when forwarding the request to the upstream_url.</Hint>
                        </Row>
                        <Row col>
                            <Label>Strips Path?</Label>
                            <Row className={b('radio-wrap')()}>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.strip_path"
                                        component={Radio}
                                        value={'true'}
                                        type="radio"
                                        id="strip-path-true"
                                    />
                                    <Label htmlFor="strip-path-true">Yes</Label>
                                </Row>
                                <Row className={b('radio')()}>
                                    <Field
                                        name="proxy.strip_path"
                                        component={Radio}
                                        value={'false'}
                                        type="radio"
                                        id="strip-path-false"
                                    />
                                    <Label htmlFor="strip-path-false">No</Label>
                                </Row>
                            </Row>
                            <Hint> Strip the path out of the listen_path when forwarding the request to the upstream_url.</Hint>
                        </Row>
                    </Row>
                </div>
                <div className={b('section')}>
                    <div className={b('section-title')}>3. Health check</div>
                    <Row className={b('row')()} fullwidth>
                        <Row col>
                            <Label>Health URL (optional)</Label>
                            <Field
                                name="health_check.url"
                                type="text"
                                placeholder="eg. http://gw.hellofresh.com/"
                                component={Input}
                            />
                            <Hint>The url that the Gateway will use to determine the health of the API. </Hint>
                        </Row>
                        <Row col>
                            <Label>Timeout (optional)</Label>
                            <Field
                                name="health_check.timeout"
                                type="number"
                                parse={parse}
                                component={Input}
                            />
                            <Hint>The length of time that the Gateway should wait before displaying an error.</Hint>
                        </Row>
                    </Row>
                </div>
                <div className={b('section')}>
                    <div className={b('section-title')}>4. Plugins</div>

                    {
                        !!plugins &&
                            <RenderPlugins
                                className={b()}
                                apiSchema={apiSchema}
                                plugins={plugins}
                                initialValues={initialValues}
                                selectedPlugins={selectedPlugins}
                                handlePluginInclude={includePlugin}
                                handlePluginExclude={removePlugin}
                            />
                    }

                */}
                </div>
            </div>
            <Row className={b('row',{ 'button-row': true })()}>
                <Button
                    type="submit"
                    mod="primary"
                >
                    Create API
                </Button>
            </Row>
        </form>
    );
};

OAuthServerForm.propTypes = propTypes;

const selector = formValueSelector('oAuthServerForm');

const form = reduxForm({
    form: 'oAuthServerForm',
    enableReinitialize: true, // this is needed!!
})(OAuthServerForm);

export default connect(
    // state => {
    //     const plugins = selector(state, 'plugins');

    //     return {
    //         keepDirtyOnReinitialize: false,
    //         plugins,
    //     };
    // },
    null,
    null,
)(form);
