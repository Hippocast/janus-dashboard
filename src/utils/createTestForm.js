import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';

const Form = ({ children, initialValues }) => children;

const createTestForm = initialValues => connect(
    () => ({
        initialValues
    }),
    null,
)(
    reduxForm({
        form: 'mockForm',
    })(Form)
);

export default createTestForm;
