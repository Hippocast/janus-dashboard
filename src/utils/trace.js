import R from 'ramda';
import { openResponseModal } from '../store/actions';

/**
 * @name trace
 * @desc helper function for development only.
 *          It should be used like this:
 * @example (from https://github.com/MostlyAdequate/mostly-adequate-guide/blob/master/ch5.md#debugging)
 *
 * import trace from '../utilz/trace';
 *
 * const dasherize = R.compose(
 *      join('-'),
 *      toLower,
 *      trace('after split'),
 *      split(' '),
 *      replace(/\s{2,}/ig, ' ')
 * );
 *
 * dasherize('The world is a vampire');
 * // after split [ 'The', 'world', 'is', 'a', 'vampire' ]
 */
const trace = R.curry((tag, x) => {
    console.error('trace -> ', tag, x);
    return x;
});

export default trace;
