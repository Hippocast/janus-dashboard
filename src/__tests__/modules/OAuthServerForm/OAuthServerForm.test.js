/* eslint-env jest */
import React from 'react'
import { createStore } from 'redux'
import { shallow } from 'enzyme'
import schema from '../../../configurations/oAuthServerSchema.config'
import { renderFakeForm } from '../../../utils/createTestForm'

import OAuthServerForm from '../../../modules/pages/NewOAuthServerPage/OAuthServerForm'

const initialValues = {
  name: 'mock-name',
  token_strategy: {
    name: 'mock-name'
  },
  oauth_endpoints: {},
  oauth_client_endpoints: {}
}

const store = createStore(() => ({
  form: {
    mockForm: {
      values: initialValues
    }
  }
}))

describe('EndpointForm component', () => {
  const requiredProps = {
    schema,
    handleSubmit: jest.fn(),
    initialValues
  }

  it('renders correctly', () => {
    const passedProps = {
      editing: false
    }
    const wrapper = shallow(
      renderFakeForm(store)(initialValues)(
        <OAuthServerForm
          {...requiredProps}
          {...passedProps}
        />
      )
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly if property `editing` is passed', () => {
    const passedProps = {
      editing: true
    }
    const wrapper = shallow(
      renderFakeForm(store)(initialValues)(
        <OAuthServerForm
          {...requiredProps}
          {...passedProps}
        />
      )
    )

    expect(wrapper).toMatchSnapshot()
  })
})
