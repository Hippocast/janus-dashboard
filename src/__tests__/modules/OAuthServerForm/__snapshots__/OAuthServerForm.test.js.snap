// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EndpointForm component renders correctly 1`] = `
<Connect(ReduxForm)>
  <Connect(ReduxForm)
    editing={false}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "name": "mock-name",
        "oauth_client_endpoints": Object {},
        "oauth_endpoints": Object {},
        "token_strategy": Object {
          "name": "mock-name",
        },
      }
    }
    schema={
      Object {
        "cors_meta": Object {
          "all_methods": Array [
            "ALL",
            "CONNECT",
            "DELETE",
            "GET",
            "HEAD",
            "OPTIONS",
            "POST",
            "PUT",
            "PATCH",
          ],
          "domains": Array [
            "*",
          ],
          "enabled": true,
          "exposed_headers": Array [
            "X-Debug-Token",
            "X-Debug-Token-Link",
          ],
          "methods": Array [],
          "request_headers": Array [
            "Origin",
            "Authorization",
            "Content-Type",
          ],
        },
        "name": "",
        "oauth_client_endpoints": Object {
          "create": Object {
            "all_methods": Array [
              "POST",
            ],
            "append_path": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "remove": Object {
            "all_methods": Array [
              "DELETE",
            ],
            "append_path": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
        },
        "oauth_endpoints": Object {
          "authorize": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "introspect": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "revoke": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "token": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
        },
        "rate_limit": Object {
          "enabled": false,
          "limit": Object {
            "labels": Array [
              "Seconds",
              "Minutes",
              "Hours",
            ],
            "unit": "",
            "units": Array [
              "S",
              "M",
              "H",
            ],
            "value": 0,
          },
        },
        "token_strategy": Object {
          "name": "",
          "settings": Array [],
          "strategies": Array [
            Object {
              "name": "jwt",
              "settings": Array [
                Object {
                  "alg": "",
                  "key": "",
                },
              ],
            },
            Object {
              "name": "introspection",
              "settings": Object {
                "auth_header_type": "",
                "use_oauth_header": false,
              },
            },
          ],
        },
      }
    }
  />
</Connect(ReduxForm)>
`;

exports[`EndpointForm component renders correctly if property \`editing\` is passed 1`] = `
<Connect(ReduxForm)>
  <Connect(ReduxForm)
    editing={true}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "name": "mock-name",
        "oauth_client_endpoints": Object {},
        "oauth_endpoints": Object {},
        "token_strategy": Object {
          "name": "mock-name",
        },
      }
    }
    schema={
      Object {
        "cors_meta": Object {
          "all_methods": Array [
            "ALL",
            "CONNECT",
            "DELETE",
            "GET",
            "HEAD",
            "OPTIONS",
            "POST",
            "PUT",
            "PATCH",
          ],
          "domains": Array [
            "*",
          ],
          "enabled": true,
          "exposed_headers": Array [
            "X-Debug-Token",
            "X-Debug-Token-Link",
          ],
          "methods": Array [],
          "request_headers": Array [
            "Origin",
            "Authorization",
            "Content-Type",
          ],
        },
        "name": "",
        "oauth_client_endpoints": Object {
          "create": Object {
            "all_methods": Array [
              "POST",
            ],
            "append_path": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "remove": Object {
            "all_methods": Array [
              "DELETE",
            ],
            "append_path": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
        },
        "oauth_endpoints": Object {
          "authorize": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "introspect": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "revoke": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
          "token": Object {
            "all_methods": Array [
              "ALL",
              "CONNECT",
              "DELETE",
              "GET",
              "HEAD",
              "OPTIONS",
              "POST",
              "PUT",
              "PATCH",
            ],
            "append_path": false,
            "enable_load_balancing": false,
            "hosts": Array [],
            "insecure_skip_verify": false,
            "listen_path": "",
            "methods": Array [],
            "preserve_host": false,
            "strip_path": false,
            "upstream_url": "",
            "upstreams": Object {
              "balancing": "",
              "options": Array [
                Object {
                  "balancing": "roundrobin",
                  "targets": Array [
                    Object {
                      "target": "",
                    },
                  ],
                },
                Object {
                  "balancing": "weight",
                  "targets": Array [
                    Object {
                      "target": "",
                      "weight": 0,
                    },
                  ],
                },
              ],
              "targets": Array [],
            },
          },
        },
        "rate_limit": Object {
          "enabled": false,
          "limit": Object {
            "labels": Array [
              "Seconds",
              "Minutes",
              "Hours",
            ],
            "unit": "",
            "units": Array [
              "S",
              "M",
              "H",
            ],
            "value": 0,
          },
        },
        "token_strategy": Object {
          "name": "",
          "settings": Array [],
          "strategies": Array [
            Object {
              "name": "jwt",
              "settings": Array [
                Object {
                  "alg": "",
                  "key": "",
                },
              ],
            },
            Object {
              "name": "introspection",
              "settings": Object {
                "auth_header_type": "",
                "use_oauth_header": false,
              },
            },
          ],
        },
      }
    }
  />
</Connect(ReduxForm)>
`;
